### PR TITLE
#24 docker (new)

### LINKED ISSUE #24: Dockerize app
- [x] Node app as docker container
- [x] connect mongoDB container
- [x] connect RabbitMQ container

### LINKED ISSUE #28: Integrate RabbitMQ
Copy-paste code from @sebastiankliem's and @mmkonrad's repository.

### PR DESCRIPTION
**Pull request by sebastiankliem**
NEW VERSION OF #41 

This dockerizes the node app and integrates configured containers for mongoDB and RabbitMQ. Development and testing using docker-compose is documented in Readme.

Could not test the AQMP/RabbitMQ integration, as there is no usable AQMP stub out there and there seems to be no way to send events from RabbitMQ in Travis.

closes #24 and closes #28

### CODE REVIEW

**Comment by mmkonrad** [comment]
👍

File: README.md [hunk 1]
```diff
@@ -8,6 +8,18 @@
...
 
+## Development with Docker
+```
+docker-compose -f docker-compose.dev.yml up
+```
+This starts the containers for the app, MongoDB and RabbitMQ. If necessary, the app's container is built automatically in advance.
+The app is then available at http://localhost:3030/.
+
+### RabbitMQ: Sending events manually
+The RabbitMQ management interface is available at http://localhost:15672. In development mode, use Username `guest` and Passowrd `guest` to login.
```
Comment (cmfcmf) [inline README.md]: typo: `username` (lowercase), `password`

File: README.md [hunk 2]
```diff
@@ -27,7 +39,7 @@ Getting up and running is as easy as 1, 2, 3.
 
 ## Testing
 
-Simply run `npm test` and all your tests in the `test/` directory will be run.
+Set the environment variable to connect to MongoDB: `MONGO_URL=mongodb://127.0.0.1:27017/gamification`. Then simply run `npm test` and all your tests in the `test/` directory will be run.
```
Comment (cmfcmf) [inline README.md]: TODO

File: Dockerfile.dev [hunk 3]
```diff
@@ -0,0 +1,8 @@
+FROM node:10.4.1
+
+COPY . /app
+WORKDIR /app
+RUN npm install -g nodemon
```
Comment (cmfcmf) [inline Dockerfile.dev]: TODO

File: .travis.yml [hunk 4]
```diff
@@ -8,6 +8,9 @@ cache:
   directories:
     - "$HOME/.npm"
 
+env:
+  - MONGO_URL=mongodb://127.0.0.1:27017/gamification
+
```
Comment (cmfcmf) [inline .travis.yml]: TODO

File: config/default.json [hunk 5]
```diff
@@ -6,5 +6,7 @@
     "default": 10,
     "max": 50
   },
-  "mongodb": "mongodb://localhost:27017/gamification"
```
Comment (cmfcmf) [inline config/default.json]: Why did you remove the url from the config file?

File: config/default.json [hunk 6]
```diff
@@ -6,5 +6,7 @@
...
+  "rabbitQueue": "my-queue",
+  "rabbitExchange": "my-exchange",
+  "rabbitRoutingKey": "testkey"
```
Comment (cmfcmf) [inline config/default.json]: Wouldn't it make sense to name this `my-key` in correspondence with the other names in this file?

File: docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  rabbitmq:
+    image: rabbitmq:3-management
+    hostname: my-rabbit
+    ports:
+      - "5672:5672"
+      - "15672:15672"
+    volumes:
+      - rabbitmq_log:/var/log/rabbitmq
+      - rabbitmq_data:/var/lib/rabbitmq/mnesia/rabbit@my-rabbit
+      - ./config/rabbitmq:/etc/rabbitmq
```
Comment (cmfcmf) [inline docker-compose.yml]: I don't think the `rabbitmq` container should be created within this file. People using our gamification service with Docker have to already have an existing RabbitMQ running somewhere. It is not our responsibility (nor should it be) to setup a `rabbitmq` container for our users.

File: get_rabbit_definitions.sh [hunk 8]
```diff
@@ -0,0 +1,6 @@
+#! /bin/bash
+
+# assumes the container to be running already
+
+docker exec -it gamification_rabbitmq_1 rabbitmqadmin -q export rabbit.definitions.json
+docker exec -it gamification_rabbitmq_1 cat rabbit.definitions.json
```
Comment (cmfcmf) [inline get_rabbit_definitions.sh]: Could you add a more detailed description to this script? Why would I want to execute it? What exactly are these definitions?

File: src/index.js [hunk 9]
```diff
@@ -8,6 +8,7 @@ process.on('unhandledRejection', (reason, p) =>
   logger.error('Unhandled Rejection at: Promise ', p, reason)
 );
 
-server.on('listening', () =>
-  logger.info('Feathers application started on http://%s:%d', app.get('host'), port)
-);
+server.on('listening', () => {
+  logger.info('Feathers application started on http://%s:%d', app.get('host'), port);
+  app.service('amqp-connector').receiveFromQueue(process.env.RABBITMQ_HOST, app.get('rabbitQueue'));
```
Comment (cmfcmf) [inline src/index.js]: Why is this an environment variable and not part of the config file?

Comment (cmfcmf) [reply src/index.js]: Also, I don't think this call really belongs to the `listening` callback. Couldn't it be placed below. the callback?

File: src/services/amqp-connector/amqp-connector.class.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    return data;
+  }
+
+  async patch(id, data, params) {
+    return data;
+  }
+
+  async remove(id, params) {
+    return { id };
+  }
```
Comment (cmfcmf) [inline src/services/amqp-connector/amqp-connector.class.js]: You made this a service, but aren

File: docker-compose.yml (hunk 7 above)
Comment (sebastiankliem) [reply docker-compose.yml]: to: docker-config for rabbit

**Comment by sebastiankliem** [comment]
all requested changes done

File: src/amqp-connector.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
+const amqp = require('amqplib');
+class AmqpConnector {
+  constructor(app) {
+    this.app = app;
+  }
+
+  async sendToQueue(host, queue, message) {
+    try {
+      let conn = await amqp.connect('amqp://' + host);
```
Comment (cmfcmf) [inline src/amqp-connector.js]: Please use `const` in this file wherever possible.

File: src/amqp-connector.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
+const amqp = require('amqplib');
+class AmqpConnector {
+  constructor(app) {
+    this.app = app;
+  }
+
+  async sendToQueue(host, queue, message) {
```
Comment (cmfcmf) [inline src/amqp-connector.js]: IMHO `host` should be passed as constructor argument, so that we don't need to re-connect every time we send something to the queue. A new method `connect` could be used to open the connection and save it in an instance variable. This method would be called just once when starting the app.

The same could be argued for the result of `createChannel()`. However, I'm no ampqlib expert, maybe channels may not be cached?

File: src/amqp-connector.js [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    }
+  }
+
+  async receiveFromQueue(host, queue) {
+    try {
+      let conn = await amqp.connect('amqp://' + host);
+      console.info('RabbitMQ connected');
+      process.once('SIGINT', () => {
+        conn.close();
+      });
```
[the remark this answers is not in the document]
Comment (sebastiankliem) [reply src/amqp-connector.js]: The connection otherwise will never be properly closed.

File: docker-compose.yml [hunk 14]
```diff
@@ -0,0 +1,27 @@
+version: '3'
+services:
+  mongo:
+    image: mongo:3.6.5
+    volumes:
+      - mongo-data:/data/db
+    ports:
+      - "27017:27017"
```
[the remark this answers is not in the document]
Comment (sebastiankliem) [reply docker-compose.yml]: OK, we can expose the port again later if needed.

**Comment by sebastiankliem** [comment]
all done.

File: Dockerfile.dev [hunk 15]
```diff
@@ -0,0 +1,7 @@
+FROM node:10.4.1
+
+COPY . /app
+WORKDIR /app
+RUN npm install
+EXPOSE 3030
+CMD ["npm","run dev"]
```
Comment (cmfcmf) [inline Dockerfile.dev]: This needs to be `CMD ["npm", "run", "dev"]`.

File: src/amqp-connector.js [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    } catch (error) {
+      console.group('RabbitMQ');
+      console.warn(error);
+      console.log('reconnecting ...');
+      console.groupEnd('rabbitmq');
+      setTimeout(() => {
+        this.connect();
+      }, 1000);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (sebastiankliem) [reply src/amqp-connector.js]: I agree. But I do not have a quick solution for that at the moment. I created an issue for that, so we can solve this later.

File: .eslintrc.json [hunk 17]
```diff
@@ -39,6 +39,7 @@
     "func-call-spacing": ["error", "never"],
     "prefer-arrow-callback": "error",
     "no-trailing-spaces": "error",
-    "comma-dangle": ["error", "never"]
+    "comma-dangle": ["error", "never"],
+    "no-console": "off"
```
Comment (frederike-ramin) [inline .eslintrc.json]: @sebastiankliem  @cmfcmf 
Actually we already got a logger library called Winston. You can see how to use it in index.js: https://github.com/frederike-ramin/gamification/blob/0857295d73aaaaf275ec036fe0f6e22bf53bbfac/src/index.js#L12

So please remove this rule and use Winston instead of console.

**Comment by cmfcmf** [comment]
Looks okay now. We'll need to add tests later on and also handle errors correctly in #48.
